### DISCUSSION TITLE
Concatenation problems in Neural Networks API

### DISCUSSION THREAD

**Discussion by AcculusSasao** [android/ndk#1801] [Q&A] [answered]
Hi, can I ask questions about NNAPI here ?

Using ANEURALNETWORKS_CONCATENATION, I found some problems.
on Snapdragon, NDK r23c.

1. Concat with axis=1 (NCHW) is not supported by DSP, and forced to be run on CPU(nnapi-reference). 
   Is it right ? No explanations are found in documents.
https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a44cbea825c4b224dd3ea757e9b1f65ed
2. Measuring time:
 for example to concatenate 2 tensors [N=1, Channels=12, Height=480, Width=640],
 concat by axis=1(NCHW) on CPU :  7 msec
 concat by axis=3(NHWC) on DSP: 94 msec
 concat by axis=3(NHWC) on CPU: 46 msec
 CPU is much faster than DSP.
  Is it right ?
 As all, how can I boost the processing time of networks including NHWC concat ?

**Comment by enh-google** [comment]
@miaowang14

**Comment by AcculusSasao** [comment]
Could you give me some kind of answer ? @miaowang14

**Comment by AcculusSasao** [comment]
Where can I ask for NNAPI ? @enh-google

**Reply by enh-google** [reply]
@miaowang14 is the person you need, not me...

**Comment by miaowang14** [comment, marked as answer]
ah, sorry to reply late. I missed the email notification.

1. NNAPI defines the operations and parameter and provides a reference CPU implementation for all the supported combination. However, the DSP hardware acceleration support are implemented by the DSP vendor. Due to hardware or software limitation, the driver may only support a subset of the ops / params, and applications needs to use ANeuralNetworksModel_getSupportedOperationsForDevices to query whether certain ops are supported or not and decide what to do.
2. In general concatenation is memory-bound, and accelerators may not be any faster than CPU. (and sometimes they are slower due to memory bandwidth limitation.

**Reply by AcculusSasao** [reply]
Thanks, we will ask to the DSP vendor.

**Comment by q-ycong-p** [comment]
Hi @miaowang14 , I am running the NNAPI demo app's [sample model](https://github.com/android/ndk-samples/blob/main/nn-samples/basic/src/main/cpp/simple_model.cpp#L113-L151) which contains Add and Mul. On a Qualcomm snapdragon chipset that has CPU, GPU and DSP, `ANeuralNetworks_getDeviceCount` and `ANeuralNetworksDevice_get*` returns 5 devices:
```
index | device type | name | version | feature-level
Device 0: NPU qti-default v1.3-20.01:build_msmnile NNAPI v30
Device 1: NPU qti-dsp v1.3-20.01:build_msmnile NNAPI v30
Device 2: GPU qti-gpu v1.3-20.01:build_msmnile NNAPI v30
Device 3: NPU google-edgetpu v1.0 NNAPI v29
Device 4: CPU nnapi-reference veng.autohu.20221028.191631 NNAPI v31
```

`ANeuralNetworksModel_getSupportedOperationsForDevices` describes the operation support on these devices as:
```
Device 0: [0, 0, 1]
Device 1: [0, 0 ,0]
Device 2: [0, 0, 1]
Device 3: [0, 0, 0]
Device 4: [1, 1, 1]
```

I am particularly interested whether NNAPI is supported by this DSP. Based on the above info, could you help me understand:
1. How can I distinguish whether Device 0 or 1 is the DSP?
2. I assume the "NNAPI v30" indicates the NPUs at least has NNAPI driver support. Is that correct?
3. Since Add and Mul are basic operations, the lack of op support for Device 0 and 1 is concerning. How can I find out WHY these basic operations are not supported on Device 0 and 1?
